src/Algebra/Graph/Relation/Symmetric.hs

```diff
@@ -11,27 +11,497 @@
 -----------------------------------------------------------------------------
 module Algebra.Graph.Relation.Symmetric (
     -- * Data structure
-    SymmetricRelation, fromRelation, toRelation,
+    SymmetricRelation(..), fromRelation, toRelation,
+
+    -- * Basic graph construction primitives
+    empty, vertex, edge, overlay, connect, vertices, edges, overlays, connects,
+
+    -- * Relations on graphs
+    isSubgraphOf,
 
     -- * Graph properties
-    neighbours
+    isEmpty, hasVertex, hasEdge, vertexCount, edgeCount, vertexList, edgeList,
+    adjacencyList, vertexSet, edgeSet, preSet, postSet, neighbours,
+
+    -- * Standard families of graphs
+    path, circuit, clique, biclique, star, stars, tree, forest,
+
+    -- * Graph transformation
+    removeVertex, removeEdge, replaceVertex, mergeVertices, gmap, induce,
+
+    -- * Relational operations
+    compose, -- closure, reflexiveClosure, symmetricClosure, transitiveClosure
   ) where
 
-import Algebra.Graph.Relation
+import qualified Algebra.Graph.Relation as R
+import qualified Algebra.Graph.Relation.Internal as RI hiding (empty, vertex, overlay, connect, referredToVertexSet)
 import Algebra.Graph.Relation.InternalDerived
 
 import qualified Data.Set as Set
+import Data.Tuple
+import Data.Tree 
+import qualified Data.Tree as Tree
 
 -- | Construct a symmetric relation from a 'Relation'.
 -- Complexity: /O(1)/ time.
-fromRelation :: Relation a -> SymmetricRelation a
+fromRelation :: R.Relation a -> SymmetricRelation a
 fromRelation = SymmetricRelation
 
 -- | Extract the underlying relation.
 -- Complexity: /O(m*log(m))/ time.
-toRelation :: Ord a => SymmetricRelation a -> Relation a
-toRelation = symmetricClosure . fromSymmetric
+toRelation :: Ord a => SymmetricRelation a -> R.Relation a
+toRelation = R.symmetricClosure . fromSymmetric
+
+-- | Construct the graph comprising /a single edge/.
+-- Complexity: /O(1)/ time, memory and size.
+--
+-- @
+-- edge x y                 == edges [(1,2),(2,1)]
+-- y)) ('connect' ('vertex' y) ('vertex' x))
+-- 'hasEdge' x y (edge x y) == True
+-- 'edgeCount'   (edge x y) <= 2
```

Comment:
This is odd. Surely, `edgeCount (edge 1 2)` should return 1, not 2.

Reply:
shouldn't the graph have the edges [(1,2), (2,1)] ?

Reply:
No, because it's the very same edge. We do not distinguish `edge 1 2` from `edge 2 1`, so reporting both of them is a duplication of information.

Reply:
What about for showing purposes, do we show all the edges ?

Reply:
Let's not show any edge duplicates.